crates/cargo-util/src/paths.rs

```diff
@@ -18,18 +18,20 @@ use tempfile::Builder as TempFileBuilder;
 /// environment variable this is will be used for, which is included in the
 /// error message.
 pub fn join_paths<T: AsRef<OsStr>>(paths: &[T], env: &str) -> Result<OsString> {
-    env::join_paths(paths.iter())
-        .with_context(|| {
-            let paths = paths.iter().map(Path::new).collect::<Vec<_>>();
-            format!("failed to join path array: {:?}", paths)
-        })
-        .with_context(|| {
-            format!(
-                "failed to join search paths together\n\
-                     Does ${} have an unterminated quote character?",
-                env
-            )
-        })
+    env::join_paths(paths.iter()).with_context(|| {
+        let paths = paths.iter()
+            .map(|s| s.as_ref().to_string_lossy())
+            .map(|s| format!("  \"{s}\""))
+            .collect::<Vec<String>>()
+            .join(",\n");
+
+        format!(
+            "failed to join paths from `${env}` together\n\
+                 If you set `${env}` manually, check if it contains an unterminated quote character \
+                 or path separators (usually `:` or `;`). Please avoid using them. \
+                 Otherwise, check if any of paths listed below contain one of those characters:\n{paths}"
+        )
+    })
```

Comment:
Thank you for the update! I've found that 

* Our error message is somewhat a bit duplicate of the one from `join_paths` ([Windows](https://github.com/rust-lang/rust/blob/85f4f41deb1557ca8ab228319d33003dd2f20f45/library/std/src/sys/windows/os.rs#L224-L228), [Unix](https://github.com/rust-lang/rust/blob/85f4f41deb1557ca8ab228319d33003dd2f20f45/library/std/src/sys/unix/os.rs#L235-L231)).
* We can write into a `mut String` instead of manually doing transform and collection.
* Perhaps use `{:?}` so we don't need to quote and escape strings by our own?

This is the current version:

```
error: failed to join paths from `$env` together
If you set `$env` manually, check if it contains an unterminated quote character or path separators (usually `:` or `;`). Please avoid using them. Otherwise, check if any o
f paths listed below contain one of those characters:
  "a:b
c",
  ""xyz",
  "abc"

Caused by:
    path segment contains separator `:`
```

Maybe we can make it less lengthy like this?

```suggestion
        let mut message = format!(
            "failed to join paths from `${env}` together\n\n\
             Check if any of path segments listed below contains \
             unterminated quote character or path separators:"
        );
        for path in paths {
            use std::fmt::Write;
            write!(&mut message, "\n    {:?}", Path::new(path)).unwrap();
        }
        message
```

Does this look a bit concise?

```
error: failed to join paths from `$env` together

Check if any of paths listed below contains unterminated quote character or path separators:
    "a:b\nc"
    "\"xyz"
    "abc"

Caused by:
    path segment contains separator `:`
```